src/TestUtilities/MessageConsumerTestCase.php

```diff
@@ -0,0 +1,158 @@
+     */
+    private $caughtException;
+
+    /**
+     * @var Exception|null
+     */
+    private $theExpectedException;
+
+    /**
+     * @var bool
+     */
+    private $assertedScenario = false;
+
+    /**
+     * @var callable
+     */
+    private $assertionCallback;
+
+    abstract public function messageConsumer(): MessageConsumer;
+
+    /**
+     * @before
+     */
+    public function setupMessageConsumer(): void
+    {
+        $this->messageConsumer = $this->messageConsumer();
+        $this->assertedScenario = false;
+    }
+
+    /**
+     * @before
+     */
+    public function unsetMessageConsumer(): void
+    {
+        unset($this->messageConsumer);
+    }
+
+    /**
+     * @return $this
+     */
+    protected function given(object ...$eventsOrMessages)
+    {
+        $this->processMessages($eventsOrMessages);
+
+        return $this;
+    }
+
+    /**
+     * @return $this
+     */
+    public function when(object ...$eventsOrMessages)
+    {
+        try {
+            $this->processMessages($eventsOrMessages);
+        } catch (Exception $exception) {
+            $this->caughtException = $exception;
+        }
+
+        return $this;
+    }
+
+    /**
+     * @return $this
+     */
+    public function expectToFail(Exception $expectedException)
+    {
+        $this->theExpectedException = $expectedException;
+
+        return $this;
+    }
+
+    /**
+     * @param object[] $eventsOrMessages
+     */
+    protected function processMessages(array $eventsOrMessages): void
+    {
+        $messages = $this->ensureEventsAreMessages($eventsOrMessages);
+
+        foreach ($messages as $message) {
+            $this->messageConsumer->handle($message);
+        }
+    }
+
+    /**
+     * @return Message[]
+     */
+    private function ensureEventsAreMessages(array $events): array
+    {
+        return array_map(function (object $event) {
+            return $event instanceof Message ? $event : new Message($event);
+        }, $events);
+    }
+
+    /**
+     * @after
+     *
+     * @throws Exception
+     */
+    protected function assertScenario(): void
+    {
+        // @codeCoverageIgnoreStart
+        if ($this->assertedScenario) {
+            return;
+        }
+        // @codeCoverageIgnoreEnd
+
+        $this->assertedScenario = true;
+
+        $this->assertExpectedException($this->theExpectedException, $this->caughtException);
+
+        if (is_callable($this->assertionCallback)) {
+            ($this->assertionCallback)($this->messageConsumer);
+        }
+    }
+
+    /**
+     * @return $this
+     */
+    public function then(callable $assertion)
+    {
+        $this->assertionCallback = $assertion;
+
+        return $this;
+    }
+
+    private function assertExpectedException(
+        Exception $expectedException = null,
+        Exception $caughtException = null
+    ): void {
+        if (null !== $caughtException && (null === $expectedException || get_class($expectedException) !== get_class(
```

Comment:
Very hard to read conditional: can it somehow be split up a bit?

Reply:
This one is horrible, I know. I've tried many times making it more readable. It kept getting uglier when trying to split it up because the parts of the condition influence each other so much. In the end I just accepted it.